File: 0021_merge_two_sorted_list/main.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{
		Val:  0,
		Next: nil,
	}
	curr := dummy
	for l1 != nil || l2 != nil {
		// to optimize, we could break the loop here when one of the node is nil
		if l1 == nil || l2 != nil && l2.Val <= l1.Val {
			curr.Next = l2
			curr = curr.Next
			l2 = l2.Next
		} else if l2 == nil || l1 != nil && l1.Val <= l2.Val {
			curr.Next = l1
			curr = curr.Next
			l1 = l1.Next
		}
	}
	return dummy.Next
}

func printList(head *ListNode) {
	if head != nil {
		fmt.Printf("%d", head.Val)
		for head.Next != nil {
			head = head.Next
			fmt.Printf(" -> %d", head.Val)
		}
	}
	fmt.Println()
}

func main() {
	l1 := &ListNode{
		Val:  1,
		Next: nil,
	}
	l2 := &ListNode{
		Val:  2,
		Next: nil,
	}
	printList(mergeTwoLists(l1, l2))
}

File: 0021_merge_two_sorted_list/main.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        head = dummy
        while l1 is not None or l2 is not None:
            if l1 is None:
                head.next = l2
                l2 = l2.next
                break
            elif l2 is None:
                head.next = l1
                l1 = l1.next
                break
            elif l1.val < l2.val:
                head.next = l1
                l1 = l1.next
            else:
                head.next = l2
                l2 = l2.next
            head = head.next
        return dummy.next